utf16fix_neon.c

```diff
@@ -6,25 +6,17 @@
 
 /*
  * Returns if a vector of type uint8x16_t is all zero.
- * Note that this only works reliably for vectors where each
- * byte is either 0 or -1.
  */
 static int
-veq_zero(uint8x16_t v)
+veq_non_zero(uint8x16_t v)
 {
-#ifdef __arm__
-	union { uint8x8_t v; double d; } narrowed;
-
-	/* narrow each byte to a nibble */
-	narrowed.v = vshrn_n_u16(vreinterpretq_u16_u8(v), 4);
-
-	return (narrowed.d == 0.0);
-#else /* AArch64 */
-	uint8x8_t narrowed;
-
-	/* check if that vector is all zero */
-	return (vdupd_lane_f64(vreinterpret_f64_u16(narrowed), 0) == 0.0);
-#endif
+	// might compile to two instructions:
+	//    umaxv   s0, v0.4s
+	//    fmov    w0, s0
+	// On Apple hardware, they both have a latency of 3 cycles, with a throughput of 
+	// four instructions per cycle. So that's 6 cycles of latency (!!!) for the two
+	// instructions. A narrowing shift has the same latency and throughput.
+	return vmaxvq_u32(vreinterpretq_u32_u8(v));
```

Comment:
This is likely slower than the other code because it has to be followed by a `tst` instruction unless `cbz` is possible; meanwhile, comparing floats sets flags directly.  This version also does not compile on armv7 due to the lack of `vmaxvq_u32`.  I do not recommend this change.

For C++ compatibility, you may use `std::bit_cast` instead of punning through a union.

Reply:
On my Apple M2 processor...

Let us call this version 'daniel':

```C++
static int
veq_non_zero(uint8x16_t v)
{
	return vmaxvq_u32(vreinterpretq_u32_u8(v));
}
```

and let us use this version as the default...

```C++
static int
veq_non_zero(uint8x16_t v)
{
	uint8x8_t narrowed = vshrn_n_u16(vreinterpretq_u16_u8(v), 4);
	return (vdupd_lane_f64(vreinterpret_f64_u16(narrowed), 0) != 0.0);
}
```

I am getting the following benchmark results:

<img width="635" alt="Screenshot 2025-01-12 at 4 16 49 PM" src="https://github.com/user-attachments/assets/b9f8b2f6-a374-44d2-be8f-9cb7f6305cd5" />

In other words, they are difficult to distinguish performance wise.

Reply:
That seems reasonable, but unfortunately won't build on armv7.  Not that much of a problem as we do not seem to support armv7-specific optimisations right now.  So yes, feel free to replace my floating-point hack with taking the maximum.  This would also avoid any issues if the caller has set flush-to-zero when calling the routines.

Note that for our string functions, we did detect a difference in performance when using the maximum vs. a floating-point hack, but it was small-ish, too.